### ISSUE TITLE
Feature: wrap and delegate 3rd party resources, to hook after or before check/in/out

### ISSUE THREAD

**Issue by keymon** [redfactorlabs/concourse-smuggler-resource#1]
# Idea

As a smuggler-resource user, I want redfactor to call any other resource bundled in the same image after or before execute my check/in/out, so that I can re-use and extend the behaviour of other resouces.
# Proposed implementation

The user can bundle any other resource binaries with the redfactor image:

```
/
└── opt
    └── resource
        ├── check
        ├── in
        ├── out
        └── wrapped-resource
            ├── ftp
            │   ├── check
            │   ├── in
            │   └── out
            ├── git
            │   ├── check
            │   ├── in
            │   └── out
            └── s3
                ├── check
                ├── in
                └── out
```

And configure smuggler like:

```
resources:
- name: tar_upload_with_license
  type: smuggler
  source:
    wrapped_resources:
    - s3:
        source:
          bucket: licensed_files_bucket
          region_name: us-east-1
          versioned_file: vpc.tfstate
          access_key_id: ...
          secret_access_key: ...
        hooks:
        - post-out
        - pre-in
        - pre-check
    smuggler_config:
      out:
        path: bash
        args:
        - -e
        - -c
        - |
          echo
          mkdir /tmp/archive
          tgzfile=${SMUGGLER_SOURCES_DIR}/${SMUGGLER_file_to_add_license}

          tar -xvf ${tgzfile} -C /tmp/archive
          echo "${SMUGGLER_license}" > /tmp/archive/LICENSE
          echo "Build on ${ATC_EXTERNAL_URL} $BUILD_JOB_NAME/$BUILD_NAME#$BUILD_ID" > /tmp/archive/BUILD_INFO
          tar -czf ${tgzfile} -C /tmp/archive

jobs:
- name: generate_upload_file
  plan:
  - task: 
    config:
      outputs:
      - name: generated-files
      # ... do somehting
  - put: tar_licenser
    params:
      license: "Blablabla legal terms" 
      file_to_add_license: generated-files/data.tgz
      wrapped_resources:
      - s3:
          file: generated-files/data.tgz
```

It will: 
- `in/check`: No config for smuggler, so it will just call s3 resource.
- `out`: Execute first a script which will add a pair of files to the tgz. Then upload the resource.
# In/Out/Check request and responses for wrapped resources

We need to provide a way to read/modify the requests&responses:
- Likely by putting the request/response in the `${SMUGGLER_INPUT_DIR}` and  `${SMUGGLER_OUTPUT_DIR}`. 
- think about formatting to make it script friendly?
# Notes
- The wrapped resource will not should know about being wrapped.  
- Of course, the image must contain all the binaries required for the wrapped resource to work. Be aware that the path of the commands would be different: `/opt/resource/wrapped-resource/s3/{in,out,check}`

**Comment by keymon** [comment]
- Q: What if the user wants to do more complex behaviour, like do some logic, call a wrapped resource, and do more logic?
- A: Maybe in that case we are going too far and a new resource shall be implemented?
- A: We can think about something like:

```
resources:
- name: tar_upload_with_license
  type: smuggler
  source:
    wrapped_resources:
    - name: s3
      type: s3 # optional
      source:
        # ...
    commands:
    - name: out-pre
      path: ...
      args: ...
    - name: out-post
      path: ...
      args: ...
    - name: in
      path: ...
      args: ...
    action_chain:
      # Execute the smuggler out-pre and out-post around the s3 out command
      out: [ smuggler:out-pre, s3:out, smuggler:out-post ]
      # This is actually the default: run the smuggler command called as the action
      in: [smuggler:in] 
```

**Comment by keymon** [comment]
Finally this is kinda implemented by #2, as:
- You can simply call any other resource directly from a script defined in smuggler.
- Smuggler will give you the Raw request as json via `Stdin`, and it can read responses in json from `Stdout`. It can even filter out any smuggler config from the request. This makes really easy to shell out to any other resource.
- If there is a requirement to modify the json of the request, then as said: the user is going too far and he must solve the issue  by his own.
